colcon_sanitizer_reports/sanitizer_log_parser.py

```diff
@@ -0,0 +1,195 @@
+                keys = []
+                break
+
+        return tuple(keys)
+
+    @staticmethod
+    def _make_default_keys(*, lines: Tuple[str]) -> Tuple[str]:
+        # Most sanitizer errors have the only/most significant stack trace first in a subsection, so
+        # we assume we've already found the correct stack trace.
+        return _KeysFactory._make_keys_base(
+            lines=lines,
+            stack_begin_regexes=(
+                r'^.*$',
+            ),
+        )
+
+    @staticmethod
+    def _make_detected_memory_leaks_keys(*, lines: Tuple[str]) -> Tuple[str]:
+        # There is one stack trace per subsection in a "detected memory leaks" error.
+        return _KeysFactory._make_keys_base(
+            lines=lines,
+            stack_begin_regexes=(
+                r'^Direct leak of X byte\(s\) in X object\(s\) allocated from:$',
+            )
+        )
+
+    @staticmethod
+    def _make_data_race_keys(*, lines: Tuple[str]) -> Tuple[str]:
+        # There are two stack traces involved a "data race" error, both in the same subsection.
+        return _KeysFactory._make_keys_base(
+            lines=lines,
+            stack_begin_regexes=(
+                r'^\s+(Read|Write) of size X at 0xX .*$',
+                r'^\s+Previous (read|write) of size X at 0xX .*$'
+            )
+        )
+
+    @staticmethod
+    def _make_lock_order_inversion_keys(*, lines: Tuple[str]) -> Tuple[str]:
+        # There are two stack traces involved in a "lock-order-inversion" error, both in the same
+        # subsection and with identical header lines.
+        return _KeysFactory._make_keys_base(
+            lines=lines,
+            stack_begin_regexes=(
+                r'^\s+Mutex M\d+ acquired here while holding mutex M\d+ in .*$',
+                r'^\s+Mutex M\d+ acquired here while holding mutex M\d+ in .*$'
+            )
+        )
+
+    @staticmethod
+    def make_keys(*, lines: Tuple[str], error_name: str) -> Tuple[str]:
+        if error_name == 'data race':
+            return _KeysFactory._make_data_race_keys(lines=lines)
+        elif error_name == 'detected memory leaks':
+            return _KeysFactory._make_detected_memory_leaks_keys(lines=lines)
+        elif error_name == 'lock-order-inversion':
+            return _KeysFactory._make_lock_order_inversion_keys(lines=lines)
+        else:
+            return _KeysFactory._make_default_keys(lines=lines)
+
+
+class _SubSection:
+    keys: Tuple[str]
+
+    def __init__(self, *, lines: Tuple[str], error_name: str) -> None:
+        self.keys = _KeysFactory.make_keys(lines=lines, error_name=error_name)
+
+
+class _Section:
+    """Contents for a single Sanitizer output section."""
+
+    error_name: str
+    sub_sections: Tuple[_SubSection]
+
+    def __init__(self, *, lines: Tuple[str]) -> None:
+        # Section name comes after 'Sanitizer: ', and before any open paren or hex number.
+        self.error_name = re.match(
```

Comment:
ditto, pre-declare the regexp, extract the regexp string at the top and document it here